opencart 1.5.6.4/upload/catalog/controller/smailyforopencart/cron_customers.php

```diff
@@ -0,0 +1,72 @@
+
+    public function index() {
+        // Customer model.
+        $this->load->model('account/customer');
+        // Load Smaily settings.
+        $this->load->model('setting/setting');
+        // Load Smaily helper.
+        $this->load->model('smailyforopencart/helper');
+        
+        // Validate token.
+
+        $settings = $this->model_setting_setting->getSetting('smaily')['smaily_customer_sync'];
+        $settings = empty($settings) ? ['enabled' => 0, 'token' => ''] : $settings;
+
+        $settings_token = isset($settings['token']) ? $settings['token'] : '';
+        $request_token = isset($this->request->get['token']) ? trim($this->request->get['token']) : '';
+        if (empty($settings_token) || empty($request_token) || $settings_token !== $request_token) {
+            die('Unauthorized');
+        }
+        if ($settings['enabled'] == 1) {
+            $offset_unsub = 0;
+            while (true) {
+                $unsubscribers = $this->model_smailyforopencart_helper->apiCall('contact', [
+                    'list' => 2,
+                    'offset' => $offset_unsub,
+                    'limit' => 2500,
+                ]);
+                // Collect unsubscriber emails.
+                $unsubscribers_emails = [];
+                foreach ($unsubscribers as $unsubscriber) {
+                    array_push($unsubscribers_emails, $unsubscriber['email']);
+                }
+                // unsubscribeCustomers method would compile a single update query.
+                $this->model_smailyforopencart_helper->unsubscribeCustomers($unsubscribers_emails);
+                $offset_unsub += 1;
+                if (empty($unsubscribers)) {
+                    break;
+                }
+            }
+            
+            $offset_sub = 0;
+            while (true) {
+                $subscribers = $this->model_smailyforopencart_helper->getSubscribedCustomers($offset_sub);
+                if (empty($subscribers)) {
+                    $this->log->write('smaily subscriber sync without customers.');
+                    echo 'No subscribers to sync.';
+                    break;
+                }
+                $list = [];
+                foreach ($subscribers as $subscriber) {
+                    // Get customer info based of selected fields from admin.
+                    $sync_fields = $this->model_smailyforopencart_helper->getSyncFields();
+                    $customer = [];
+                    foreach ($sync_fields as $field) {
+                        $customer[$field] = $subscriber[$field];
+                        $offset_sub = $subscriber['customer_id'];
+                    }
+                    $customer['is_unsubscribed'] = "0";
+                    array_push($list, $customer);
+                }
+                // Send subscribers to smaily.
+                $response = $this->model_smailyforopencart_helper->apiCall('contact', $list, 'POST');
```

Comment:
What do you have in mind, doesn't the apiCall function have error handling?
I currently added a log->write to line 69 with json_encode($response)

Reply:
Yes, `apiCall` has error handling, but it just logs the error. 

Bigger issue is that customers synchronization would just move on to the next chunk. A better solution would be for the CRON task to exit with non-zero exit code in case of an API request failure. That way system administrators can monitor if CRON task is running correctly or not.